Refactor TokenUpdater to use Ticker

Coffeeri opened the pull request:
Depends on #36

Coffeeri committed:
refactor: improved error handling

Coffeeri committed:
ci: add golangci-lint and use go version 1.X

Coffeeri committed:
style: rename hTTPClient

Coffeeri committed:
refactor(httpx): use ticker for token updater

Coffeeri committed:
feature(httpx): httpclient interface for testing

adankb commented on selling_partner.go:
```diff
@@ -36,12 +36,13 @@ type Client struct {
 }
 
 func NewClient(config Config) (*Client, error) {
+	hc := http.DefaultClient
```
Der HTTP-Client sollte zur Config hinzugefügt werden, damit der User der Library einen eigenen HTTP-Client übergeben kann. Das sollte aber optional sein. Wenn der `nil` ist, sollte die Lib einen eigenen HTTP-Client erstellen.

adankb commented on selling_partner.go:
```diff
@@ -26,7 +26,7 @@ type Config struct {
 
 type Client struct {
-	httpClient  *httpx.Client
+	hTTPClient  *httpx.Client
```
Das sieht komisch aus 🙂
Entweder "httpClient" oder "HTTPClient" würde ich sagen. Oder meckert der Linter dann?

adankb commented on apis/caller.go:
```diff
@@ -82,19 +81,26 @@ func (a *Call[responseType]) Execute(httpClient HTTPClient) (*CallResponse[respo
-	if a.ParseErrorListOnError && !callResp.IsSuccess() {
-		if err = unmarshalBody(resp, &callResp.ErrorList); err != nil {
-			return nil, err
+	if callResp.IsSuccess() { // Response is OK
```
Im Function Body sollten wir nur Kommentare schreiben, wenn es absolut notwendig ist, weil man ohne den Kommentar die Funktion schlechter verstehen würde.
"`if callResp.IsSuccess()`" ist ja selbsterklärend.

adankb commented on apis/caller.go:
```diff
@@ -82,19 +81,26 @@ func (a *Call[responseType]) Execute(httpClient HTTPClient) (*CallResponse[respo
-		if err = unmarshalBody(resp, &callResp.ResponseBody); err != nil {
-			return nil, err
+
+	// Response is not OK
```
Würde den Kommentar hier auch weglassen.

adankb commented on httpx/token_updater.go:
```diff
@@ -46,76 +41,103 @@ type AccessTokenResponse struct {
+					t.log.Errorf("Failed to fetch new access-tokenAPI: %s", err.Error())
+					continue
+				}
+				t.accessToken.Store(&token.AccessToken) // store new token
```
Würde den Kommentar hier auch weglassen.

adankb commented on httpx/token_updater.go:
```diff
@@ -46,76 +41,103 @@ type AccessTokenResponse struct {
+				t.accessToken.Store(&token.AccessToken) // store new token
+				secondsToWait := token.ExpiresIn - int(constants.ExpiryDelta/time.Second)
+				durationToWait := time.Duration(secondsToWait) * time.Second
+				ticker.Reset(durationToWait) // reset ticker for next token request
```
Würde den Kommentar hier auch weglassen.

adankb commented on httpx/client.go:
```diff
@@ -26,37 +26,49 @@ type ClientConfig struct {
+	awsSession             *session.Session
+}
+
+type HTTPRequester interface {
```
Brauchen wir hier wirklich ein eigenes Interface?
Würde ansonsten http.Client vorziehen. Dann weiß man direkt, was das ist.

adankb commented on httpx/client_test.go:
```diff
@@ -58,7 +56,7 @@ func Test_httpClient_addAccessToken(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			h := &Client{
-				httpClient:   tt.fields.HTTPClient,
+				hTTPClient:   tt.fields.HTTPClient,
```
Der `HTTPClient` kann aus der `fields` struct entfernt werden - er ist immer `nil`.

adankb commented on httpx/token_updater.go:
```diff
@@ -46,76 +41,103 @@ type AccessTokenResponse struct {
-	})
+// RunInBackground starts a goroutine that fetches a new access token periodically
+// and stores it in the client. The goroutine is stopped when the returned cancel function is called.
+func (t *PeriodicTokenUpdater) RunInBackground() (cancel func(), err error) {
```
`RunInBackground()` führt den ersten TokenRequest nicht mehr direkt synchron aus.
Ich denke, dass wir dieses Verhalten nicht ändern sollten, damit es direkt knallt, wenn z. B. die Config nicht stimmt. Sonst sieht man die Fehler nur im Log.

adankb commented on httpx/token_updater.go:
```diff
@@ -46,76 +41,103 @@ type AccessTokenResponse struct {
+	done := make(chan bool)
+
+	go func() {
+		for {
```
Cool! Durch den ticker und die for-Schleife wird es auf jeden Fall klarer, wie der token-Updater funktioniert 🙂

adankb commented on httpx/token_updater.go:
```diff
@@ -46,76 +41,103 @@ type AccessTokenResponse struct {
+			case <-ticker.C:
+				token, err := t.doTokenRequest()
+				if err != nil {
+					t.log.Errorf("Failed to fetch new access-tokenAPI: %s", err.Error())
```
Im Fehlerfall sollte der Ticker auch noch mal gestartet werden.

Coffeeri replied:
Der Ticker läuft ewig weiter, bis er gestoppt wird.
Hier habe ich aber nochmal einen Default `constants.DefaultTokenUpdaterBackoffTime` eingebaut der im ständigen Fehlerfall gewartet wird.

adankb reviewed (commented):
Hi Leander, nicht erschrecken, aber mir sind ein paar Sachen aufgefallen 🙂

Coffeeri committed:
feature(token-updater): Add initial token fetch

This  value is used to set the first ticker duration.
If the request fails the ticker will be set to `DefaultTokenUpdaterBackoffTime`.

Coffeeri committed:
feature(config): Add optional HTTPClient

Coffeeri committed:
style: Rename hTTPClient

Coffeeri committed:
refactor(httpx-test): Remove unneeded HTTPClient

It is always set to nil

Coffeeri committed:
refactor(api-caller): Make error path more clear

Coffeeri committed:
style(api-caller): Remove obvious comment

Coffeeri committed:
feat(orders): add Approvals, ShipmentConfirmation

Add new Amazon API changes of the orders-api and minor refactorings.

Changelog:

* fix: installation url in README.md

* feat: add Set in internal utils pkg

* feat: add approvals and shipmentConfirmation to OrdersAPI

* feat: add new types in orders model.go

* refactor: use enums in orders api with sets

* fix(orders)!: add missing callResponse return
  Applies to
  + `UpdateOrderItemsApprovals`
  + `ConfirmShipment`

VloRyan committed:
Fix: parse rdt error list (#39)

* fix: parsing of errorList at RDT
---------

Co-authored-by: Florian <florian.meiberg@fondof.de>

adankb committed:
chore: rebase

adankb committed:
fix: linter warnings

adankb committed:
fix: merge
